modules/cli/src/main/scala/scala/cli/commands/scalafix/Scalafix.scala

```diff
@@ -0,0 +1,197 @@
+      .withScalaVersion(scalaVersion)
+
+    logger.debug(
+      s"Processing ${sourcePaths.size} Scala sources"
+    )
+
+    val rulesThatWillRun: Either[ScalafixException, mutable.Buffer[ScalafixRule]] =
+      try
+        Right(scalafix.rulesThatWillRun().asScala)
+      catch
+        case e: ScalafixException => Left(e)
+    val needToBuild: Boolean = rulesThatWillRun match
+      case Right(rules) => rules.exists(_.kind().isSemantic)
+      case Left(_)      => true
+
+    val preparedScalafixInstance = if (needToBuild) {
+      val res = Build.build(
+        inputs,
+        buildOptionsWithSemanticDb,
+        compilerMaker,
+        None,
+        logger,
+        crossBuilds = false,
+        buildTests = false,
+        partial = None,
+        actionableDiagnostics = actionableDiagnostics
+      )
+      val builds = res.orExit(logger)
+
+      val successfulBuildOpt = for {
+        build <- builds.get(Scope.Main)
+        sOpt  <- build.successfulOpt
+      } yield sOpt
+
+      val classPaths = successfulBuildOpt.map(_.fullClassPath).getOrElse(Seq.empty)
+      val externalDeps =
+        options.shared.dependencies.compileOnlyDependency ++ successfulBuildOpt.map(
+          _.options.classPathOptions.extraCompileOnlyDependencies.values.flatten.map(_.value.render)
+        ).getOrElse(Seq.empty)
+      val scalacOptions = options.shared.scalac.scalacOption ++ successfulBuildOpt.map(
+        _.options.scalaOptions.scalacOptions.toSeq.map(_.value.value)
+      ).getOrElse(Seq.empty)
+
+      scalafix
+        .withScalacOptions(scalacOptions.asJava)
+        .withClasspath(classPaths.map(_.toNIO).asJava)
+        .withToolClasspath(Seq.empty.asJava, externalDeps.asJava)
+    }
+    else
+      scalafix
+
+    val customScalafixInstance = preparedScalafixInstance
+      .withParsedArguments(options.scalafixArg.asJava)
+
+    val errors = if (options.check) {
+      val evaluation = customScalafixInstance.evaluate()
+      if (evaluation.isSuccessful)
+        evaluation.getFileEvaluations.foldLeft(List.empty[String]) {
+          case (errors, fileEvaluation) =>
+            val problemMessage = fileEvaluation.getErrorMessage.toScala.orElse(
+              fileEvaluation.previewPatchesAsUnifiedDiff.toScala
+            )
+            errors ++ problemMessage
+        }
+      else
+        evaluation.getErrorMessage.toScala.toList
+    }
+    else
+      customScalafixInstance.run().map(prepareErrorMessage).toList
+
+    if (errors.isEmpty) sys.exit(0)
+    else {
+      errors.tapEach(logger.error)
+      sys.exit(1)
+    }
+  }
+
+  private def prepareErrorMessage(error: ScalafixError): String = error match
+    case ParseError => "A source file failed to be parsed"
+    case CommandLineError =>
+      "A command-line argument was parsed incorrectly"
+    case MissingSemanticdbError =>
+      "A semantic rewrite was run on a source file that has no associated META-INF/semanticdb/.../*.semanticdb"
```

Comment:
nit: all scalafix rules are not rewrites
```suggestion
      "A semantic rule was run on a source file that has no associated META-INF/semanticdb/.../*.semanticdb"
```

Reply:
removed this code - it isn't needed after switching on launcher